Stream open/queuing redux

jagerman opened the pull request:
This PR addresses some issues with btrequest error/timeout flags; and some issues with how streams get constructed and queued with some conflicts when trying to mix queue_stream, get_new_stream, and a stream constructor callback.

(This is based on #71; only the last two commits here are relevant to this PR)

As for the stream changes:

There was a confusing behaviour with `queue_stream` and `get_new_stream` that, even if given an explicit stream type and arguments, those arguments and type would be ignored if there was a stream constructor callback and so you really had to go *only* stream constructor callback *or* get_new_stream/queue_stream because mixing them would break in unexpected ways.

This addresses the issue (along with related changes along the way):

- rename `queue_stream` to `queue_incoming_stream` to better reflect what it does.
- rename `get_new_stream` to `open_stream` to better reflect what it does.
- split `queue_incoming_stream` and `open_stream` into two versions:
    - an untemplated version that takes no arguments and always just constructs a default stream (firing the stream constructor callback, if set; making a basic Stream otherwise).
    - a templated version that always bypasses the stream constructor callback and always creates the requested type with the given arguments.
- Added a `maybe_stream` alternative to `get_stream` that returns nullptr if the stream id doesn't exist, and throws if the stream exists but isn't compatible with the given templated type (if given a Stream subclass to cast to).
- Changed `get_stream` to throw a different exception (invalid_argument) if the stream exists but has an incompatible type, and changed its stream-does-not-exist exception to a std::out_of_range.
- added a virtual `on_ready()` function to Stream that subclasses can override if they want to do something when the stream becomes ready.
- various test cases for the new stream queuing/opening/retrieving that adds various additional tests for edge cases and mixed queuing/constructor callback usage.

jagerman committed:
Change `listen` return to `void`

It never returns false: it only ever throws, or returns true, so convert
it to void with a void promise so now it either throws on failure, or
doesn't throw on success.

jagerman committed:
Move direction methods to connection_interface

So that it is accessible when you have the generic one, e.g. from
pybind.

jagerman committed:
Remove dubious double-enable_shared_from_this

The one in connection_interface is used, but the second one in
Connection doesn't appear to be, and seems potentially dangerous: we
have two different shared_pointers inside an Connection instance and it
isn't clear that they are aware of each other.

Removing it affected nothing else, so seems fine to drop.

jagerman committed:
Make `send(bstring)` less dangerous

Currently we don't allow calling `send` with a std::string or ustring,
but a bstring slips through the cracks via the base method
implementation: calling `send(my_bstring)` ends up calling effectively
`send(bstring_view{my_bstring})`, but this ends up sending garbage
because that `my_bstring` likely doesn't live long enough.

This rearranges the virtual method to be protected so that bstring gets
the same protection as std::string and ustring, avoiding an error-prone
case.

(I encountered this in innocuous-looking code that was doing exactly
this, and ending up sending garbage on the wire).

*With* this change, `send(my_bstring)` no longer compiles, forcing you
to change it to `send(std::move(my_bstring))` or else doing something
with a keep-alive, as appropriate.

jagerman committed:
Don't fail for sending an empty string

There's nothing specifically *wrong* with sending an empty string: it
just doesn't do anything (just like you can print an empty string to a
file or stdout and have it do nothing).

This was throwing (unexpected) exceptions when I casually sent some data
that happened to be empty, and it felt weird to have to make the
application do a `if (!empty)` everywhere it sends something.

jagerman committed:
Format

jagerman committed:
Add stream buffer ownership transfer test case

I wrote this up while investigating a crash on the Python wrapper side.
This passes just fine here (and so it's a python wrapper issue), but
having written it it seems a worthwhile test to have.

(We have other tests that explicitly rely on these things, but not one
that definitely tests just this one issue).

jagerman committed:
Hold dgram_interface in a shared_ptr

This aligns it with how Stream works, and makes it much easier to use
ith pybind (where pybind wants to be able to share ownership via a
shared_ptr holder).

dr7ana committed:
linter, ci

jagerman committed:
Fix race condition in stream()

The weak pointer could away between the call to `expired()` and the
subsequent `lock()`.  `lock()`ing first, then checking that lock
returned non-empty avoids the race.

jagerman committed:
Update oxen-encoding (for bstring support)

jagerman committed:
Fail for bad input on request length parser

Currently the logic here continually accumulates a buffer until a `:` is
found, but someone could abuse that by sending a very large stream of
data that never contains a `:` to run the receiver out of memory.  This
change puts a maximum on the length parser to mitigate this.

(There was also a bug here that was checking `.back()` instead of
`npos`)

jagerman committed:
Convert btstream to use more bstring and bstring_views

There were some unnecessary copies in here, but also using a bstring
matches much better what we're actually sending on the wire,
particularly when interacting with Python.  The internals now use
bstring/bstring_views and the public-facing APIs now take the body as
either a regular string_view or a bstring_view.

jagerman committed:
Fix issues with multiple sequential bt requests

Adds a test that blasts 50 requests at once, which triggered some
failures (fixed here):

- the request id lower_bound lookup had a reversed condition and, as a
  result, didn't properly find requests when there were more than one in
  the container.

- The `buf` we build up to pass to handle_input had `current_len` added
  to it, but wasn't taking into account that we might already have some
  of the request loaded into `buf` from a previous request (and so we
  could end up losing some of the data needed for processing the next
  request).

dr7ana committed:
Address:set_port

jagerman committed:
Add some tests of big request success and bigger request failures

jagerman committed:
Fix class/struct mismatch

jagerman committed:
Error handling cleanup

This refactors the handling of error codes to address a few
shortcomings and problems:

- error codes were kind of scattered all over, but in terms of numeric
  value, and in terms of where they are defined.  This concentrates them
  both numerically and in the code (now in error.hpp).
- error codes are now all values above 777'000'000, which makes them
  easier to read out of logs (e.g. 777000100: 777 means a libquic error,
  and 100 indicates stream exception).  Previously we were using values
  under 2<<62, but those seem pointlessly huge (and aren't trivially
  easy to decipher out of a log).
- the error code for an expiring connection was invalid (2<<62 + 1,
  which is larger than the max allowed QUIC integer), and so when a
  connection timed out an ngtcp2 assert failed and aborted the program.
- fixed some weirdness in stream.close() that took an io_error and sent
  an *ngtcp2* error code, but that wasn't right because those are
  internal, negative codes: we need to always send a uint64_t
  application error code.  This changes close to simply take an error
  code.
- added a custom exception class (`application_stream_error`) that
  application code can throw to fire a custom error code down the
  stream (this solves a TODO in the code).
- redid quic_strerror to return string that can also display a custom
  error code, if used.
- removed the `error::` enum because it felt kind of wrong to be
  static_cast'ing uint64_t values into the error:: enum for enum values
  that it didn't actually hold.  (This is valid C++, but feels a bit
  dirty).

jagerman committed:
Simplify make_command/make_response

They were doing a bit too much: opt:: values got passed through, as well
as rid incrementing and exception handling, plus the actual encoding.
This renames them to `encode_command()`/`encode_response()` and has them
only do the encoding part, leaving the rest to the existing caller which
splits up the logic involved fairly cleanly.

jagerman committed:
Misc cleanups

- change stream count variables to uint64_t; having them as int's
  introduces a bunch of pain when interacting with the ngtcp2
  input/outputs that are uint64_t.
- remove redundant BTRequestStream destructor (it was doing part of what
  post-destructor destruction has to do anyway).
- address a bunch of clang-tidy warnings about using reserved
  double-underscore prefixes in variables

jagerman committed:
Debug-log connection destruction in debug builds

jagerman committed:
Fix idle close crash

By not causing `set_closing()` for the ERR_IDLE_CLOSE case we weren't
safe against repeated closes, and so if a connection close happened
during stream destruction after an idle close (as lokinet was doing)
we'd double-close and crash.

jagerman committed:
Make public close_connection interface thread-safe

Lokinet was (innocently) calling this from another thread, but that can
and will break things.

Also fixes a potential crash with `close_conns` that could suffer from
iterator invalidation because of close_connection() changing `conns`
while it iterates.

jagerman committed:
Defer libquic object destructions to the event loop

This adds a `.make_shared<T>` to Network and Endpoint that works almost
like std::make_shared<T>, except that it ensures that the destructor
will be invoked in the Network's event loop, which is not necessarily
the thread that releases the last shared_ptr.

This is implemented by a two-pass custom deleter: when the last
shared_ptr is released, rather than immediately triggering destruction
(as in a standard shared_ptr) we instead queue the raw pointer
destruction onto the event loop thread to perform the destruction.

This commit also makes the stream, btstream, and datagram constructors
protected, with Network as a friend class, so that they *have* to go
through net.make_shared (because std::make_shared will produce a
shared_ptr that destructs in the wrong place).

jagerman committed:
Allow stream_construct_cb to return nullptr

If it returns nullptr we construct a default stream.  This makes the
callback a little more flexible: it doesn't have to worry about how a
default stream should be constructed, but can only worry about special
streams.

dr7ana committed:
expose remote pubkey from connection interface

dr7ana committed:
expose get_stream from connection level

dr7ana committed:
view remote_key method tweak

dr7ana committed:
templated stream getter is better

jagerman committed:
Fix `message` move/copy constructors

Moving sometimes, and copying always, would leave you with a `data`
member with different memory from before, but all the string_view's
pointing into the *old* data member.

This adds explicit copy/move ctors/assignments that properly update all
the string_views to point into the new member.

jagerman committed:
Fix segv in check_timeout

`f` never got updated to point to the new first element after erasing.

jagerman committed:
(squash me) message: nicer solution to copy/move issues

Updating the string_views was a bit hairy when, essentially, what we
need are offset/sizes of the substr within `data`.  The changes the
implementation to just store offset/size pairs, rather than
string_views, which simplifies things a lot as we can now just use the
default copy/move constructors again.

jagerman committed:
Fix wrong type in enable_if for operator=

jagerman committed:
Fix crash in register_command

If not in the logic thread, the deferred call here was keeping invalid
references.

jagerman committed:
Fix from_chars not checking that we consumed the whole string

As written it would accept something like "123abc:" as length 123 and
just ignore the "abc", but we should be failing in such a case.

jagerman committed:
Timeout & error handling fixes

- Fix dangling reference in timeout handling
- Fix shadowing in `message` constructor that resulted in the `is_error`
  member never getting set.
- Rename `is_error` constructor parameter to `is_timeout` to reflect
  what it actually sets (and fix the previous point).
- make `operator bool` explicit to avoid foot shotguns
- update `operator bool` comment to mention is_error
- use the time_point typedef more

dr7ana committed:
Stream construction, message timeouts?

- We need the stream id to be passed in stream construction to specify btreq streams
- message timeout vs error needs to be re-granulared

dr7ana committed:
testing something

dr7ana committed:
stream count method

jagerman committed:
Revamp stream construction/queuing

There was a confusing behaviour with `queue_stream` and `get_new_stream`
that, even if given an explicit stream type and arguments, those
arguments and type would be ignored if there was a stream constructor
callback.

This addresses it along with related changes along the way:

- rename `queue_stream` to `queue_incoming_stream` to better reflect
  what it does.
- rename `get_new_stream` to `open_stream` to better reflect what it
  does.
- split `queue_incoming_stream` and `open_stream` into two versions:
    - an untemplated version that takes no arguments and always just
      constructs a default stream (firing the stream constructor, if
      set), or making a basic Stream otherwise.
    - a templated version that always bypasses the stream constructor
      callback and always creates the requested type with the given
      arguments.
- Added a `maybe_stream` alternative to `get_stream` that returns
  nullptr if the stream id doesn't exist, and throws if the stream
  exists but isn't compatible with the given templated type (if given a
  Stream subclass to cast to).
- Changed `get_stream` to throw a different exception (invalid_argument)
  if the stream exists but has an incompatible type, and changed its
  stream-does-not-exist exception to a std::out_of_range.
- added a virtual `on_ready()` function to Stream that subclasses can
  override if they want to do something when the stream becomes ready.
- various test cases for the new stream queuing/opening/retrieving that
  adds various additional tests for edge cases and mixed
  queuing/constructor callback usage.

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -55,48 +56,97 @@ namespace oxen::quic
+        /// not currently an open stream; throws std::invalid_argument if the stream exists but is
+        /// not an instance of the given StreamT type.
+        template <typename StreamT = Stream, std::enable_if_t<std::is_base_of_v<Stream, StreamT>, int> = 0>
+        std::shared_ptr<StreamT> maybe_stream(int64_t id)
```
thanks for adding this!

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -55,48 +56,97 @@ namespace oxen::quic
+            if constexpr (!std::is_same_v<StreamT, Stream>) {
+                if (auto st = std::dynamic_pointer_cast<StreamT>(std::move(s)))
+                    return st;
+                throw std::invalid_argument{"Stream ID {} is not an instance of the requested Steam subclass"_format(id)};
```
Small typo ("Steam")

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -55,48 +56,97 @@ namespace oxen::quic
+        virtual std::shared_ptr<Stream> queue_incoming_stream_impl(
                 std::function<std::shared_ptr<Stream>(Connection& c, Endpoint& e)> make_stream) = 0;
-        virtual std::shared_ptr<Stream> get_new_stream_impl(
+        virtual std::shared_ptr<Stream> open_stream_impl(
```
I think this re-naming is a lot clearer on intent and purpose

dr7ana commented on include/quic/btstream.hpp:
```diff
@@ -36,7 +36,7 @@ namespace oxen::quic
+        message(BTRequestStream& bp, bstring req, bool is_timeout = false);
 
       public:
         void respond(bstring_view body, bool error = false);
```
Discussion point:
Currently our response types are "errors" ("E") and "responses" ("R"), but we could introduce a "timeout" ("T") option to actually make a useful differentiation between the two in the application code. Otherwise, we'd have to use some standardized "timeout response" that could be parsed by the library code

jagerman replied:
Timeout is not a response; it happens because we didn't get *any* response to the request within a certain amount of time.

dr7ana commented on src/connection.cpp:
```diff
@@ -373,25 +373,31 @@ namespace oxen::quic
-            auto stream = construct_stream(make_stream);
+            std::shared_ptr<Stream> stream;
+            if (make_stream)
+                stream = make_stream(*this, _endpoint);
```
Why do you only pass the optional ID to the make_stream in `Connection::stream_opened`? This completely eliminates the use case of queueing streams to be opened in the client's on_conn_open callback, for example

dr7ana committed:
test cases, stream queueing

- Added 2 test cases, one for server stream constructor callback use, the other for server stream queueing in the connection open callback
- small change to stream getter to check stream queue for the second use case

dr7ana commented:
I added a few test cases: 
The first test case extracts streams created in the server stream constructor and confirming ID's and such. If using the stream constructor, the server cannot use `maybe_stream` to get the btstream until the client has sent a request down

The other was an issue i noticed in stream queueing for inbound connections. If a server queue's a btreq stream to be opened in its connection open constructor (like in lokinet), it cannot use the `::maybe_stream` method to get it until the client sends something down that stream, upon which it is moved from `stream_queue` to `streams`. I made a quick change to `Connection::get_stream_impl` to check `stream_queue` after checking `streams,` which passed the test case.

Though they went from all failing from 002 (from the connection line that prints the huge requests) to failing in the 004 test cases prior to the ones i added i think... i haven't been able to reproduce locally either which is weird

dr7ana committed:
configurable keep-alive timeout

dr7ana committed:
oxenc vbump

jagerman committed:
Make sure there is no garbage after bt request stream data

jagerman committed:
Disable over-logging of big request tests

The test here produces 500k lines of debug logging when at debug log
level, so force it up to (at least) info for the duration of the test.

dr7ana committed:
CI fixes
